### PR TITLE
PWGHF: add Lc ML scores in Sc output.

### PR DESCRIPTION
**Pull request by mfaggin**

### CODE REVIEW

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 1]
```diff
@@ -392,9 +437,16 @@ struct HfTaskSigmac {
...
         if (enableTHn) {
-          /// fill it only if no MC operations are enabled, otherwise fill it in the processMC with the right origin and channel!
-          if (!doprocessMc) {
-            registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, 0, 0, ptSc, std::abs(chargeSc));
+          if (!isMc) {
+            /// fill it only if no MC operations are enabled, otherwise fill it in the processMC with the right origin and channel!
+            if constexpr (useMl) {
+              /// fill with ML information
+              /// BDT index 0: bkg score; BDT index 2: non-prompt score
+              registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, candidateLc.mlProbLcToPiKP()[0], candidateLc.mlProbLcToPiKP()[2], 0, 0, ptSc, std::abs(chargeSc));
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:445]: The vector `mlProbLcToPiKP` is filled (i.e. with non-zero size) only for `Lc` candidates that fulfil all the preselections. However if I am not wrong here you could have Lc candidates that didn't pass the preselections depending on the value of this configurable https://github.com/AliceO2Group/O2Physics/blob/f64af44dd7a98be8782a9dda39e86b5731c0483d/PWGHF/TableProducer/candidateCreatorSigmac0plusplus.cxx#L49 set in the Sigmac creator. 

Wouldn't be better to check the size or the Lc flag before filling it in the THnSparse given that there is this possibility, what do you think?

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 2]
```diff
@@ -412,27 +464,42 @@ struct HfTaskSigmac {
...
+              registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, 0, 0);
+            }
           }
           if (candidateLc.isSelLcToPiKP() >= 1) {
             massLc = hfHelper.invMassLcToPiKP(candidateLc);
-            registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, 0, 0);
+            if constexpr (useMl) {
+              /// fill with ML information
+              /// BDT index 0: bkg score; BDT index 2: non-prompt score
+              registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, candidateLc.mlProbLcToPiKP()[0], candidateLc.mlProbLcToPiKP()[2], 0, 0);
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:481]: Same as above

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 3]
```diff
@@ -564,7 +631,7 @@ struct HfTaskSigmac {
 
       /// get the candidate Λc+ used to build the Σc0
       /// and understand which mass hypotheses are possible
-      const auto& candidateLc = candSc.prongLc_as<soa::Join<aod::HfCand3Prong, aod::HfSelLc, aod::HfCand3ProngMcRec>>();
+      const auto& candidateLc = candSc.prongLc_as<L>();
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:634]:
```suggestion
      const auto& candidateLc = candSc.prongLc_as<CandsLc>();
```

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 4]
```diff
@@ -412,27 +464,42 @@ struct HfTaskSigmac {
...
-                 soa::Join<aod::McParticles, aod::HfCand3ProngMcGen> const& mcParticlesLc,
-                 aod::McParticles const& mcParticles, // this establishes the type of particle obtained with the .mcParticle() getter
-                 soa::Join<aod::HfCand3Prong, aod::HfSelLc, aod::HfCand3ProngMcRec> const& candidatesLc,
-                 aod::TracksWMc const&)
+  template <bool useMl, typename L>
+  void fillHistosMc(soa::Join<aod::HfCandSc, aod::HfCandScMcRec> const& candidatesSc,
+                    soa::Join<aod::McParticles, aod::HfCandScMcGen> const& mcParticlesSc,
+                    soa::Join<aod::McParticles, aod::HfCand3ProngMcGen> const& mcParticlesLc,
+                    aod::McParticles const& mcParticles, // this establishes the type of particle obtained with the .mcParticle() getter
+                    L const& candidatesLc,
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:501]:
```suggestion
  template <bool useMl, typename CandsLc>
  void fillHistosMc(soa::Join<aod::HfCandSc, aod::HfCandScMcRec> const& candidatesSc,
                    soa::Join<aod::McParticles, aod::HfCandScMcGen> const& mcParticlesSc,
                    soa::Join<aod::McParticles, aod::HfCand3ProngMcGen> const& mcParticlesLc,
                    aod::McParticles const& mcParticles, // this establishes the type of particle obtained with the .mcParticle() getter
                    CandsLc const& candidatesLc,
```

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 5]
```diff
@@ -886,23 +981,78 @@ struct HfTaskSigmac {
...
+            registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, origin, channel);
+          }
         }
         if (candidateLc.isSelLcToPiKP() >= 1 && pdgAbs == kPiPlus) {
           massLc = hfHelper.invMassLcToPiKP(candidateLc);
-          registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, origin, channel);
+          if constexpr (useMl) {
+            /// fill with ML information
+            /// BDT index 0: bkg score; BDT index 2: non-prompt score
+            registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, candidateLc.mlProbLcToPiKP()[0], candidateLc.mlProbLcToPiKP()[2], origin, channel);
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:1003]: Same as above

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 6]
```diff
@@ -886,23 +981,78 @@ struct HfTaskSigmac {
...
+        if (candidateLc.template prong0_as<aod::TracksWMc>().has_mcParticle()) {
+          pdgAbs = std::abs(candidateLc.template prong0_as<aod::TracksWMc>().mcParticle().pdgCode());
         }
         if (candidateLc.isSelLcToPKPi() >= 1 && pdgAbs == kProton) {
           massLc = hfHelper.invMassLcToPKPi(candidateLc);
-          registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, origin, channel);
+          if constexpr (useMl) {
+            /// fill with ML information
+            /// BDT index 0: bkg score; BDT index 2: non-prompt score
+            registry.get<THnSparse>(HIST("hnLambdaC"))->Fill(ptLc, massLc, candidateLc.mlProbLcToPKPi()[0], candidateLc.mlProbLcToPKPi()[2], origin, channel);
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:992]: Same as above

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 7]
```diff
@@ -864,7 +952,14 @@ struct HfTaskSigmac {
 
           /// THn for candidate Σc0,++ cut variation
           if (enableTHn) {
-            registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, origin, channel, ptSc, std::abs(chargeSc));
+            if constexpr (useMl) {
+              /// fill with ML information
+              /// BDT index 0: bkg score; BDT index 2: non-prompt score
+              registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, candidateLc.mlProbLcToPiKP()[0], candidateLc.mlProbLcToPiKP()[2], origin, channel, ptSc, std::abs(chargeSc));
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:958]: Same as above

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 8]
```diff
@@ -804,12 +885,19 @@ struct HfTaskSigmac {
 
           /// THn for candidate Σc0,++ cut variation
           if (enableTHn) {
-            registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, origin, channel, ptSc, std::abs(chargeSc));
+            if constexpr (useMl) {
+              /// fill with ML information
+              /// BDT index 0: bkg score; BDT index 2: non-prompt score
+              registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, candidateLc.mlProbLcToPKPi()[0], candidateLc.mlProbLcToPKPi()[2], origin, channel, ptSc, std::abs(chargeSc));
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:891]: Same as above

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 9]
```diff
@@ -715,7 +789,14 @@ struct HfTaskSigmac {
 
           /// THn for candidate Σc0,++ cut variation
           if (enableTHn) {
-            registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, origin, channel, ptSc, std::abs(chargeSc));
+            if constexpr (useMl) {
+              /// fill with ML information
+              /// BDT index 0: bkg score; BDT index 2: non-prompt score
+              registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, candidateLc.mlProbLcToPiKP()[0], candidateLc.mlProbLcToPiKP()[2], origin, channel, ptSc, std::abs(chargeSc));
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:795]: Same as above

File: PWGHF/D2H/Tasks/taskSigmac.cxx [hunk 10]
```diff
@@ -653,12 +720,19 @@ struct HfTaskSigmac {
 
           /// THn for candidate Σc0,++ cut variation
           if (enableTHn) {
-            registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, decLengthLc, decLengthXYLc, cpaLc, cpaXYLc, origin, channel, ptSc, std::abs(chargeSc));
+            if constexpr (useMl) {
+              /// fill with ML information
+              /// BDT index 0: bkg score; BDT index 2: non-prompt score
+              registry.get<THnSparse>(HIST("hnSigmaC"))->Fill(ptLc, deltaMass, candidateLc.mlProbLcToPKPi()[0], candidateLc.mlProbLcToPKPi()[2], origin, channel, ptSc, std::abs(chargeSc));
```
Comment (fgrosa) [inline PWGHF/D2H/Tasks/taskSigmac.cxx:726]: Same as above

**Review by fgrosa — CHANGES_REQUESTED** [review]
Hi @mfaggin, apart from a minor suggestion, I have a doubt about the vector of ML scores. Can you please check? Thanks!

File: PWGHF/D2H/Tasks/taskSigmac.cxx (hunk 1 above)
Comment (mfaggin) [reply PWGHF/D2H/Tasks/taskSigmac.cxx:445]: You are totally right, indeed filling with something empty (as done in the selector) != not filling.
Thanks for spotting it, I'll solve it

**Comment by mfaggin** [comment]
> Hi @mfaggin, apart from a minor suggestion, I have a doubt about the vector of ML scores. Can you please check? Thanks!

thanks @fgrosa indeed you are right: in the selector we always fill with BDT score vectors, which however might be empty is the candidate do not pass all preselections.
I have fixed it here, thanks

**Review by fgrosa — APPROVED** [review]

**Pull request merged**